File: src/Cop.java
import java.util.ArrayList;
import java.util.Random;

/**
 * Simulates a Cop.
 */
class Cop extends Turtle {
    /**
     * Construct a new Cop object.
     * @param world the world this cop is in
     */
    Cop(World world) {
        super(world);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void update() {
        super.update();
        enforce();
    }

    /**
     * Cop can always move.
     */
    @Override
    boolean canMove() {
        return true;
    }

    /**
     * Find and arrest a random active agent in the neighbourhood
     */
    private void enforce() {
        // Find all active agents in the neighbourhood
        ArrayList<Turtle> agents = world.getPatchManager().filterNeighbourTurtle(patch, (turtle ->
            turtle instanceof Agent && ((Agent) turtle).isActive()
        ));

        // Dont continue if there is no matched agent
        if (agents.size() == 0) {
            return;
        }

        // Select a random one from the matched agents
        Random r = new Random();
        int index = r.nextInt(agents.size());
        Agent suspect = (Agent) agents.get(index);

        // Move to the patch of the (about-to-be) jailed agent
        moveToPatch(suspect.getPatch());

        // Arrest suspect
        suspect.setActive(false);
        suspect.setJailTerm(r.nextInt(Params.MAX_JAILED_TERM + 1));
    }

    /**
     * Returns a symbol representing cop.
     * @return symbol in string
     */
    @Override
    public String toString() {
        return "C";
    }
}

File: rebellion-python/models/Cop.py
import random

from models import Turtle, World, Agent
from params import MAX_JAILED_TERM

# Author: Dafu Ai


class Cop(Turtle):
    """
    Simulates a Cop.
    """
    def __init__(self, world: World) -> None:
        super().__init__(world)

    def update(self) -> None:
        """Perform relevant action as a Cop."""
        super().update()
        self.enforce()

    def can_move(self) -> bool:
        """Cops can always move if parent class allows movement."""
        return super().can_move()

    def enforce(self) -> None:
        """Find and arrest a random active agent in the neighbourhood."""

        # Find all active agents in the neighbourhood
        agents = self.world.patch_map.filter_neighbour_turtles(
            self.patch,
            lambda t: isinstance(t, Agent) and t.active
        )

        # Don't continue if there is no matched agent
        if len(agents) == 0:
            return

        # Move to the patch of the (about-to-be) jailed agent
        suspect = random.choice(agents)
        self.move_to_patch(suspect.patch)

        # Arrest suspect
        suspect.active = False
        suspect.jail_term = random.randint(1, MAX_JAILED_TERM)



